### PR TITLE
End-to-End HTTPS enablement

### PR DESCRIPTION
**Pull request by DukeDavis12**

### CODE REVIEW

**Comment by Darshini-Parikh93** [comment]
Let me know if my understanding is correct:
1) We have added truststore in owner to perform TO0 in HTTPS mode
2) At a time a service can be run either in HTTP or HTTPS mode but not in both

**Comment by Darshini-Parikh93** [comment]
Please update README of each service for this feature

**Comment by DukeDavis12** [comment]
> Let me know if my understanding is correct:
> 
> 1. We have added truststore in owner to perform TO0 in HTTPS mode
    -- Yes.. For RV certificate verification during TO0.
> 2. At a time a service can be run either in HTTP or HTTPS mode but not in both
   -- Yes. Either HTTP or HTTPS.

File: demo/device/README.md [hunk 1]
```diff
@@ -72,4 +72,25 @@ $ mvn -Dfido.iot.pem.dev=<device-PEM-file> exec:java
...
+
+  ```
+    java -D<other-flags> -Dfido_ssl_mode=TEST device.jar
+  ```
+
+  Make sure to add the `-Dfido_ssl_mode=TEST`. Default SSL mode is `TEST`.
+
+  * `PROD` mode where certificate verification is carried out. Useful for production deployment.
+  ```
+    java -D<other-flags> -Dfido_ssl_mode=PROD -Dssl_trustore=<trust-store-path> -Dssl_truststore-password=<truststore-pass> -Dssl_truststore_type=<truststore-type> -jar device.jar
```
Comment (Sai-Anudeep47) [inline demo/device/README.md:93]: Update flag ssl keystore password to -Dssl_truststore_password

File: demo/owner/README.md [hunk 2]
```diff
@@ -180,4 +180,32 @@ The PKCS12 keystore file \<fido-iot-src\>/demo/owner/owner_keystore.p12 contains
...
+By default, the Owner uses HTTP for all communications on port 8042. In addition to that, the Owner can be configured to handle HTTPS request from the device.
+
+- Generate the Keystore/Certificate for the Owner
+
+  * Ensure that the web certificate is issued to the resolvable domain of the Owner server.
+
+- Copy the generated Keystore/Certificate to `demo/owner/certs` folder.
+
+- Copy the truststore containing all the required certificates to `demo/owner/certs` folder.
+
```
Comment (Sai-Anudeep47) [inline demo/owner/README.md:196]: Add instructions to update RV blob for HTTPS

File: demo/device/README.md (hunk 1 above)
Comment (DukeDavis12) [reply demo/device/README.md]: Done.

File: demo/owner/README.md (hunk 2 above)
Comment (DukeDavis12) [reply demo/owner/README.md:196]: Done.

File: demo/rv/README.md [hunk 3]
```diff
@@ -123,4 +123,27 @@ $ openssl x509 -in cert.pem -pubkey -noout | openssl enc -base64 -d > example_al
...
+
+- Copy the generated Keystore/Certificate to `demo/rv/certs` folder.
+
+- Update the following environment varibles in `demo/rv/rv.env` file
+
+    |  Variable            |  Value            |             description       |
+    | ---------------------|-------------------|-------------------------------|
+    | rv_protocol_scheme  | https             | To enable HTTPS communication.|
+    | rv_https_port       | port number       | The given port number will be used for HTTPS
+    communication. |
```
Comment (Darshini-Parikh93) [inline demo/rv/README.md:144]: `communication` is rendering into a separate row.

File: demo/rv/README.md [hunk 4]
```diff
@@ -123,4 +123,27 @@ $ openssl x509 -in cert.pem -pubkey -noout | openssl enc -base64 -d > example_al
...
+    | rv_ssl_keystore     | keystore-filename | filename of Keystore that is present in the certs folder.|
+    | rv_ssl_keystore-password| keystore-password | password of the keystore. |
+
+    **NOTE:** Appropriate security measures with respect to key-store management should be considered while performing production deployment of Manufacturer.
```
Comment (Darshini-Parikh93) [inline demo/rv/README.md:148]: Manufacturer -> RV

File: demo/owner/README.md [hunk 5]
```diff
@@ -180,4 +180,33 @@ The PKCS12 keystore file \<fido-iot-src\>/demo/owner/owner_keystore.p12 contains
...
+    | owner_https_port       | port number       | The given port will be used for HTTPS communication. |
+    | fido_ssl_mode          | TEST / PROD       | If set to `TEST`, then SSL verification is disabled. If set to `PROD`, then certificate verification is initiated. |
+    | owner_ssl_keystore     | keystore-filename | Filename of Keystore that is present in the certs folder.|
+    | owner_ssl_keystore-password| keystore-password | Password of the keystore. |
+    | ssl_truststore         | truststore-filename  | Filename of truststore that is present in the certs folder. Not required in `TEST` mode. |
+    | ssl_truststore_password| truststore-password | Password of the truststore. Not required in `TEST` mode. |
+    | ssl_truststore_type    | truststore-type   | Type of truststore. eg: JKS ,PKCS12   |
+    | owner_to0_rv_blob      | to0_rv_blob       | Contains the to0_rv_blob used by device to connect with the Owner during T02. Eg: https://localhost:<owner-https-port>?ipaddress=127.0.0.1 |
+
+    **NOTE:** Appropriate security measures with respect to key-store management should be considered while performing production deployment of Manufacturer.
```
Comment (Darshini-Parikh93) [inline demo/owner/README.md:211]: Manufacturer -> Owner

File: demo/owner/README.md [hunk 6]
```diff
@@ -180,4 +180,33 @@ The PKCS12 keystore file \<fido-iot-src\>/demo/owner/owner_keystore.p12 contains
...
+    | -----------------------|-------------------|-------------------------------|
+    | owner_protocol_scheme  | https             | To enable HTTPS communication.|
+    | owner_https_port       | port number       | The given port will be used for HTTPS communication. |
+    | fido_ssl_mode          | TEST / PROD       | If set to `TEST`, then SSL verification is disabled. If set to `PROD`, then certificate verification is initiated. |
+    | owner_ssl_keystore     | keystore-filename | Filename of Keystore that is present in the certs folder.|
+    | owner_ssl_keystore-password| keystore-password | Password of the keystore. |
+    | ssl_truststore         | truststore-filename  | Filename of truststore that is present in the certs folder. Not required in `TEST` mode. |
+    | ssl_truststore_password| truststore-password | Password of the truststore. Not required in `TEST` mode. |
+    | ssl_truststore_type    | truststore-type   | Type of truststore. eg: JKS ,PKCS12   |
+    | owner_to0_rv_blob      | to0_rv_blob       | Contains the to0_rv_blob used by device to connect with the Owner during T02. Eg: https://localhost:<owner-https-port>?ipaddress=127.0.0.1 |
```
Comment (Darshini-Parikh93) [inline demo/owner/README.md:209]: "\<owner-https-port\>" is not rendering. Add "\\" before the opening and ending angular braces.

File: component-samples/owner/src/main/java/org/fido/iot/sample/OwnerContextListener.java [hunk 7]
```diff
@@ -169,7 +169,7 @@ protected void failed(Exception e) {
     // create tables
     OwnerDbManager manager = new OwnerDbManager();
     manager.createTables(ds);
-    manager.importVoucher(ds, Composite.fromObject(VOUCHER));
+    //manager.importVoucher(ds, Composite.fromObject(VOUCHER));
```
[the remark this answers is not in the document]
Comment (DukeDavis12) [reply component-samples/owner/src/main/java/org/fido/iot/sample/OwnerContextListener.java:172]: as HTTPS is mandated for T00, the default voucher resolves directives for T00 to https://localhost:8040 thus resulting in an error.

File: component-samples/rv/src/main/java/org/fido/iot/sample/RvServerApp.java [hunk 8]
```diff
@@ -19,6 +21,15 @@
           ? Integer.parseInt(RvConfigLoader.loadConfig(RvAppSettings.TO0_TO1_PORT))
           : 8040;
 
+  private static final int RV_HTTPS_PORT =
+      null != RvConfigLoader.loadConfig(RvAppSettings.RV_HTTPS_PORT)
+          ? Integer.parseInt(RvConfigLoader.loadConfig(RvAppSettings.RV_HTTPS_PORT))
+          : 8443;
```
[the remark this answers is not in the document]
Comment (DukeDavis12) [reply component-samples/rv/src/main/java/org/fido/iot/sample/RvServerApp.java:27]: The suggested value was 8443.

File: util/http-dispatchers/http-client-dispatcher/src/main/java/org/fido/iot/sample/WebClient.java [hunk 9]
```diff
@@ -68,6 +80,83 @@ protected SSLParameters getSslParematers() {
...
+        };
+
+        SSLContext sslContext = SSLContext.getInstance("SSL");
+        sslContext.init(null, trustAllCerts, new java.security.SecureRandom());
+        return sslContext;
+      } else if (sslMode.toLowerCase().equals("prod")) {
+
+        String trustStoreFile = System.getProperty("ssl_truststore","truststore");
+        String trustStorePwd = System.getProperty("ssl_truststore_password"," ");
+        String trustStoreType = System.getProperty("ssl_truststore_type","PKCS12");
```
[the remark this answers is not in the document]
Comment (DukeDavis12) [reply util/http-dispatchers/http-client-dispatcher/src/main/java/org/fido/iot/sample/WebClient.java:118]: Using the truststore naming convention from iot-platform-sdk.

File: demo/owner/README.md [hunk 10]
```diff
@@ -180,4 +180,32 @@ The PKCS12 keystore file \<fido-iot-src\>/demo/owner/owner_keystore.p12 contains
...
-As the H2 DB grows, larger heap space will be required by the application to run the service. Default configured heap size is `256 MB`. Increase the heap size appropriately in `demo/owner/owner-entrypoint.sh` to avoid heap size issues.
\ No newline at end of file
+As the H2 DB grows, larger heap space will be required by the application to run the service. Default configured heap size is `256 MB`. Increase the heap size appropriately in `demo/owner/owner-entrypoint.sh` to avoid heap size issues.
+
+# Configuring Owner for HTTPS/TLS Communication
+
+By default, the Owner uses HTTP for all communications on port 8042. In addition to that, the Owner can be configured to handle HTTPS request from the device.
+
+- Generate the Keystore/Certificate for the Owner
+
```
[the remark this answers is not in the document]
Comment (DukeDavis12) [reply demo/owner/README.md:190]: Done

File: demo/manufacturer/README.md [hunk 11]
```diff
@@ -111,4 +111,26 @@ Refer the [Docker Commands](../README.md/#docker-commands) to start the service.
...
 
-**IMPORTANT** This is an example implementation using simplified credentials. This must be changed while performing production deployment
\ No newline at end of file
+**IMPORTANT** This is an example implementation using simplified credentials. This must be changed while performing production deployment
+
+# Configuring Manufacturer for HTTPS/TLS Communication
+
+By default, the PRI-Manufacturer uses HTTP for all communications on port 8039. In addition to that, the PRI-Manufacturer can be configured to handle HTTPS request from the device.
+
+- Generate the Keystore/Certificate for the PRI-manufacturer
```
[the remark this answers is not in the document]
Comment (DukeDavis12) [reply demo/manufacturer/README.md]: Done.

File: demo/rv/README.md [hunk 12]
```diff
@@ -123,4 +123,27 @@ $ openssl x509 -in cert.pem -pubkey -noout | openssl enc -base64 -d > example_al
...
 
-***NOTE:*** Input file cert.pem is an X509 certificate.
\ No newline at end of file
+***NOTE:*** Input file cert.pem is an X509 certificate.
+
+# Configuring RV for HTTPS/TLS Communication
+
+By default, the RV uses HTTP for all communications on port 8040. In addition to that, the RV can be configured to handle HTTPS request from the owner & device on port 8443.
+
+- Generate the Keystore/Certificate for the RV
```
[the remark this answers is not in the document]
Comment (DukeDavis12) [reply demo/rv/README.md]: Done.

File: demo/rv/README.md (hunk 3 above)
Comment (DukeDavis12) [reply demo/rv/README.md]: Done.

File: demo/rv/README.md (hunk 4 above)
Comment (DukeDavis12) [reply demo/rv/README.md]: Done.

File: demo/owner/README.md (hunk 5 above)
Comment (DukeDavis12) [reply demo/owner/README.md]: Done.

File: demo/owner/README.md (hunk 6 above)
Comment (DukeDavis12) [reply demo/owner/README.md]: Done.

**Review by Darshini-Parikh93 — APPROVED** [review]

**Review by johneasterday — APPROVED** [review]

**Pull request merged**